granulate_utils/metrics/yarn/utils.py

```diff
@@ -0,0 +1,273 @@
+import logging
+import os
+import re
+import socket
+import subprocess
+import xml.etree.ElementTree as ET
+from dataclasses import dataclass
+from functools import cached_property
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from granulate_utils.config_feeder.core.utils import mask_sensitive_value
+
+REGEX_YARN_VAR = re.compile(r"\${([^}]+)}")
+
+RM_HIGH_AVAILABILITY_ENABLED_PROPERTY_KEY = "yarn.resourcemanager.ha.enabled"
+RM_HIGH_AVAILABILITY_IDS_PROPERTY_KEY = "yarn.resourcemanager.ha.rm-ids"
+RM_HOSTNAME_PROPERTY_KEY = "yarn.resourcemanager.hostname"
+RM_WEB_ADDRESS_PROPERTY_KEY = "yarn.resourcemanager.webapp.address"
+RM_DEFAULTS = {
+    RM_HOSTNAME_PROPERTY_KEY: "0.0.0.0",
+    RM_WEB_ADDRESS_PROPERTY_KEY: "${yarn.resourcemanager.hostname}:8088",
+    RM_HIGH_AVAILABILITY_ENABLED_PROPERTY_KEY: "false",
+}
+
+RM_DEFAULT_ADDRESS = "http://0.0.0.0:8088"
+WORKER_ADDRESS = "http://0.0.0.0:8042"
+
+
+class YarnConfigError(Exception):
+    pass
+
+
+@dataclass(frozen=True)
+class YarnNodeInfo:
+    """YARN node information
+
+    Args:
+        config: YARN config from yarn-site.xml.
+
+        resource_manager_webapp_addresses:
+          When high-availability mode is enabled list of all ResourceManager webapp addresses,
+          otherwise list with single RM address.
+
+        resource_manager_index:
+          Set only if node is a ResourceManager.
+    """
+
+    config: Dict[str, str]
+    resource_manager_webapp_addresses: List[str]
+    resource_manager_index: Optional[int] = None
+
+    @cached_property
+    def is_resource_manager(self) -> bool:
+        return self.resource_manager_index is not None
+
+    @cached_property
+    def is_first_resource_manager(self) -> bool:
+        return self.resource_manager_index == 0
+
+
+def get_yarn_node_info(
+    *, logger: Union[logging.Logger, logging.LoggerAdapter], yarn_config: Optional[Dict[str, str]] = None
+) -> Optional[YarnNodeInfo]:
+    """
+    If running on YARN return YARN node information
+    """
+    config = detect_yarn_config(logger=logger) if yarn_config is None else yarn_config
+    if config is None:
+        return None
+    if rm_addresses := get_resource_manager_addresses(config, logger=logger):
+        return YarnNodeInfo(
+            resource_manager_index=get_rm_index(rm_addresses, logger=logger),
+            resource_manager_webapp_addresses=rm_addresses,
+            config=config,
+        )
+    return None
+
+
+def get_rm_index(
+    rm_addresses: List[str],
+    *,
+    logger: Union[logging.Logger, logging.LoggerAdapter],
+    hostname: Optional[str] = None,
+    ip: Optional[str] = None,
```

Comment:
hostname and ip are never used in function calls?

Reply:
fixed

Reply:
Ok, but now `get_yarn_node_info` is never called with hostname and IP?

Reply:
fixed